include/pmacc/cuSTL/algorithm/kernel/Foreach.hpp

```diff
@@ -92,11 +92,55 @@ struct Foreach
     BOOST_PP_REPEAT_FROM_TO(1, BOOST_PP_INC(FOREACH_KERNEL_MAX_PARAMS), FOREACH_OPERATOR, _)
 };
 
+
 #undef FOREACH_OPERATOR
 #undef SHIFT_CURSOR_ZONE
 #undef SHIFTED_CURSOR
 
+template<
+    uint32_t T_numWorkers,
+    typename BlockDim
+>
+struct ForeachLockstep
+{
+
+    /* operator()(zone, functor, cursor0, cursor1, ..., cursorN-1)
+     *
+     * \param zone Accepts currently only a zone::SphericZone object (e.g. containerObj.zone())
+     * \param functor or lambdaFun either a functor with N arguments or a N-ary lambda function (e.g. _1 = _2)
+     * \param args cursor for the N-th data source (e.g. containerObj.origin())
+     *
+     * The functor is called for each worker within the zone.
+     * It is called like functor(*cursor0(cellBlockOffset), ..., *cursorN(cellBlockOffset))
+     *
+     */
+    template<
+        typename T_Zone,
```

Comment:
instead of `T_Zone` let us use `T_SphericZone` with it's `uint` signature as a free template to enforce that only spheric zones (see docstring) can be used.